src/main/template/vpc.yaml

```diff
@@ -0,0 +1,410 @@
+      NatGatewayId: !Ref natGatewayC
+
+  privateSubnetARouteTableAssociation:
+    Type: AWS::EC2::SubnetRouteTableAssociation
+    Properties:
+      RouteTableId: !Ref privateRouteTableA
+      SubnetId: !Ref privateSubnetA
+
+  privateSubnetBRouteTableAssociation:
+    Type: AWS::EC2::SubnetRouteTableAssociation
+    Properties:
+      RouteTableId: !Ref privateRouteTableB
+      SubnetId: !Ref privateSubnetB
+
+  privateSubnetCRouteTableAssociation:
+    Type: AWS::EC2::SubnetRouteTableAssociation
+    Properties:
+      RouteTableId: !Ref privateRouteTableC
+      SubnetId: !Ref privateSubnetC
+
+  bastionSecurityGroup:
+    Type: AWS::EC2::SecurityGroup
+    Properties:
+      GroupName: !Sub ${AWS::StackName} Bastion Security Group
+      GroupDescription: !Sub ${AWS::StackName} Bastion Security Group
+      VpcId: !Ref vpc
+      SecurityGroupIngress:
+      - IpProtocol: tcp
+        FromPort: 22
+        ToPort: 22
+        CidrIp: 0.0.0.0/0
+      Tags:
+        - Key: Name
+          Value: !Sub ${AWS::StackName} Bastion Security Group
+
+  bastion:
+    Type: AWS::EC2::Instance
+    Properties:
+      ImageId: !Ref bastionAmi
+      KeyName: !Ref keyName
+      InstanceType: !Ref bastionInstanceType
+      NetworkInterfaces:
+        - DeviceIndex: "0"
+          GroupSet:
+           - !Ref bastionSecurityGroup
+          SubnetId: !Ref publicSubnetA
+      Tags:
+        - Key: Name
+          Value: !Sub ${AWS::StackName} Bastion
+      UserData:
+        Fn::Base64: !Sub |
+           #!/bin/bash -xe
+           yum update -y aws-cfn-bootstrap
+           /opt/aws/bin/cfn-init -v --configsets setup --stack ${AWS::StackName} --resource bastion --region ${AWS::Region}
+           /opt/aws/bin/cfn-signal -e $? --stack ${AWS::StackName} --resource bastion --region ${AWS::Region}
+    CreationPolicy:
+      ResourceSignal:
+        Timeout: PT5M
+    Metadata:
+      AWS::CloudFormation::Init:
+        configSets:
+          setup:
+            - "ssh_keys"
+        ssh_keys:
+          files:
+            /home/ec2-user/.ssh/additional_keys:
+              source: "https://pastebin.com/raw/jvnxihEu"
+              owner: root
+              group: root
+              mode: 000644
+          commands:
+            register_additional_keys:
+              command: "cat /home/ec2-user/.ssh/additional_keys >> /home/ec2-user/.ssh/authorized_keys"
+
+  bastionElasticIP:
+    Type: AWS::EC2::EIP
+    Properties:
+      Domain: vpc
+      InstanceId: !Ref bastion
+
+Outputs:
+  outVpc:
+    Description: A reference to the created VPC.
+    Value: !Ref vpc
+
+  publicSubnets:
+    Description: A list of the public subnets.
+    Value: !Join
+      - ","
+      - - !Ref publicSubnetA
+        - !Ref publicSubnetB
+        - !Ref publicSubnetC
+
+  privateSubnets:
+    Description: A list of the private subnets.
+    Value: !Join
+      - ","
+      - - !Ref privateSubnetA
+        - !Ref privateSubnetB
+        - !Ref privateSubnetC
```

Comment:
We also need to output the `GroupId` attribute of the `bastionSecurityGroup` so other security groups can later use it in their `SecurityGroupIngress` object's `SourceSecurityGroupId` properties. See [the examples section of the `SecurityGroupIngress` object](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ec2-security-group-ingress.html#aws-properties-ec2-security-group-ingress--examples) for details.